GitExtUtils/GitUI/ThreadHelper.cs

```diff
@@ -0,0 +1,122 @@
+using System.Threading.Tasks;
+using System.Windows.Forms;
+using Microsoft.VisualStudio.Threading;
+
+namespace GitUI
+{
+    public static class ThreadHelper
+    {
+#pragma warning disable SA1139 // Use literal suffix notation instead of casting
+        private const int RPC_E_WRONG_THREAD = unchecked((int)0x8001010E);
+#pragma warning restore SA1139 // Use literal suffix notation instead of casting
+
+        private static JoinableTaskContext _joinableTaskContext;
+        private static JoinableTaskCollection _joinableTaskCollection;
+        private static JoinableTaskFactory _joinableTaskFactory;
+
+        public static JoinableTaskContext JoinableTaskContext
+        {
+            get
+            {
+                return _joinableTaskContext;
+            }
+
+            internal set
+            {
+                if (value == _joinableTaskContext)
+                {
+                    return;
+                }
+
+                if (value == null)
+                {
+                    _joinableTaskContext = null;
+                    _joinableTaskCollection = null;
+                    _joinableTaskFactory = null;
+                }
+                else
+                {
+                    _joinableTaskContext = value;
+                    _joinableTaskCollection = value.CreateCollection();
+                    _joinableTaskFactory = value.CreateFactory(_joinableTaskCollection);
+                }
+            }
+        }
+
+        public static JoinableTaskFactory JoinableTaskFactory => _joinableTaskFactory;
+
+        public static void ThrowIfNotOnUIThread([CallerMemberName] string callerMemberName = "")
+        {
+            if (!JoinableTaskContext.IsOnMainThread)
+            {
+                string message = string.Format(CultureInfo.CurrentCulture, "{0} must be called on the UI thread.", callerMemberName);
+                throw new COMException(message, RPC_E_WRONG_THREAD);
+            }
+        }
+
+        public static void ThrowIfOnUIThread([CallerMemberName] string callerMemberName = "")
+        {
+            if (JoinableTaskContext.IsOnMainThread)
+            {
+                string message = string.Format(CultureInfo.CurrentCulture, "{0} must be called on a background thread.", callerMemberName);
+                throw new COMException(message, RPC_E_WRONG_THREAD);
+            }
+        }
+
+        public static void FileAndForget(this JoinableTask joinableTask, Func<Exception, bool> fileOnlyIf = null)
+        {
+            joinableTask.Task.FileAndForget(fileOnlyIf);
+        }
+
+        public static void FileAndForget(this Task task, Func<Exception, bool> fileOnlyIf = null)
+        {
+            JoinableTaskFactory.RunAsync(
+                async () =>
+                {
+                    try
+                    {
+                        await task.ConfigureAwait(false);
+                    }
+                    catch (Exception ex) when (fileOnlyIf?.Invoke(ex) ?? true)
```

Comment:
Sure enough, FileAndForget propagates OperationCancelledException  to the UI thread.
After https://github.com/gitextensions/gitextensions/commit/66d13c4cff34eb1ff50aa3287a6d68ded15af963 GE crashes when InvokeAsync is called on a disposed control.

Reply:
Do you have a stack trace for one such exception? I'm trying to write a test case for it.

Reply:
I created a failing test: https://github.com/gitextensions/gitextensions/commit/688f235a48fae4121ec215ab0c5e773381dec85d

Reply:
The test fails when run in separation but does not fail when run all tests is performed.

Reply:
This is a big help, thanks!

❓ Should we _always_ ignore `OperationCanceledException`, or should we only ignore it when `fileOnlyIf` is null?

Reply:
It should always be ignored. If it is needed to catch OperationCanceledException it can always be done with a dedicated try catch block around the awaiting statement.

Reply:
Great! I'm just finishing up a big group of tests for the different ways it can be used.